functional_tests.go

```diff
@@ -121,6 +121,37 @@ func ignoredLog(function string, args map[string]interface{}, startTime time.Tim
 	return log.WithFields(fields)
 }
 
+// Delete objects and buckets recursively
+func cleanupBuckets(c *minio.Client) error {
+	var err error
+	var buckets []minio.BucketInfo
+
+	// Create a done channel to control 'ListObjectsV2' go routine.
+	doneCh := make(chan struct{})
+
+	// Indicate to our routine to exit cleanly upon return.
+	defer close(doneCh)
+
+	buckets, err = c.ListBuckets()
+
+	// Iterate over each bucket, list objects in the bucket and delete
+	for _, bucket := range buckets {
+		go func(bucketName string) {
+			objectCh := c.ListObjectsV2(bucketName, "", true, doneCh)
+			for object := range objectCh {
+				if object.Err != nil {
+					if object.Key != "" {
+						err = c.RemoveObject(bucketName, object.Key)
+					}
+				} else {
+					err = object.Err
```

Comment:
Do not set the err from outside scope in a go-routine,  collect all errors.. in a errs []error and then store them based on the buckets index. like this 

```go
for i, bucket := range buckets {
     go func(i int, bucketName string) {
           ....
           errs[i] = err
     }(i, bucket)
}
```

Reply:
Also since there is no wait group, this won't delete all buckets all the time. Since the program might exit if this function returns after all go-routines have spawned. 

Use wait group to limit yourself and use `wg.Wait()` to wait for cleaning up of all buckets. Once done loop through 

```go
for _, err := range errs {
        if err != nil {
             return err 
        }
}
```